File: c/src/tui/terminal.c
#include "terminal.h"
#include "color.h"
#include <stdio.h>
#include <sys/ioctl.h>
#include <unistd.h>

void terminal_get_size(int *width, int *height) {
    struct winsize ws;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &ws);
    *width = ws.ws_col;
    *height = ws.ws_row;
}

void terminal_enter_raw(Terminal *term) {
    struct termios raw;
    tcgetattr(STDIN_FILENO, &term->orig);
    raw = term->orig;
    raw.c_lflag &= ~(ECHO | ICANON | ISIG);  // no echo, canonical mode, or signals
    raw.c_iflag &= ~(IXON | ICRNL);          // no flow control, no CR->NL
    raw.c_cc[VMIN] = 0;                       // non-blocking read
    raw.c_cc[VTIME] = 0;
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw);
}

void terminal_exit_raw(Terminal *term) {
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &term->orig);
}

void terminal_setup(Terminal *term) {
    terminal_enter_raw(term);
    printf(ESC_ALT_SCREEN_ON ESC_CURSOR_HIDE ESC_MOUSE_ON ESC_CLEAR_SCREEN);
    fflush(stdout);
}

void terminal_restore(Terminal *term) {
    printf(ESC_RESET_ATTRS ESC_MOUSE_OFF ESC_CURSOR_SHOW ESC_ALT_SCREEN_OFF);
    fflush(stdout);
    terminal_exit_raw(term);
}

File: rust/src/tui/terminal.rs
//! Terminal RAII wrapper using libc termios

use super::escape;
use std::io::{self, Write};
use std::mem::MaybeUninit;

pub struct Terminal {
    orig_termios: libc::termios,
    initialized: bool,
}

impl Terminal {
    pub fn new() -> io::Result<Self> {
        let mut orig_termios = MaybeUninit::<libc::termios>::uninit();

        // Get original terminal settings
        if unsafe { libc::tcgetattr(libc::STDIN_FILENO, orig_termios.as_mut_ptr()) } != 0 {
            return Err(io::Error::last_os_error());
        }

        let orig_termios = unsafe { orig_termios.assume_init() };
        let mut raw = orig_termios;

        // Set raw mode flags (matching C/C++/Java implementations)
        // Disable echo, canonical mode, signals, extended input processing
        raw.c_lflag &= !(libc::ECHO | libc::ICANON | libc::ISIG | libc::IEXTEN);

        // Disable XON/XOFF flow control, CR to NL mapping
        raw.c_iflag &= !(libc::IXON | libc::ICRNL);

        // Disable output processing
        raw.c_oflag &= !libc::OPOST;

        // Set 8-bit chars
        raw.c_cflag |= libc::CS8;

        // Non-blocking read: VMIN = 0, VTIME = 0
        raw.c_cc[libc::VMIN] = 0;
        raw.c_cc[libc::VTIME] = 0;

        if unsafe { libc::tcsetattr(libc::STDIN_FILENO, libc::TCSAFLUSH, &raw) } != 0 {
            return Err(io::Error::last_os_error());
        }

        // Setup alternate screen, cursor, mouse
        print!(
            "{}{}{}{}",
            escape::ALT_SCREEN_ON,
            escape::CURSOR_HIDE,
            escape::MOUSE_ON,
            escape::CLEAR_SCREEN
        );
        io::stdout().flush()?;

        Ok(Terminal {
            orig_termios,
            initialized: true,
        })
    }

    pub fn size() -> io::Result<(u16, u16)> {
        let mut ws = MaybeUninit::<libc::winsize>::uninit();
        if unsafe { libc::ioctl(libc::STDOUT_FILENO, libc::TIOCGWINSZ, ws.as_mut_ptr()) } != 0 {
            return Ok((80, 24)); // Fallback
        }
        let ws = unsafe { ws.assume_init() };
        Ok((ws.ws_col, ws.ws_row))
    }
}

impl Drop for Terminal {
    fn drop(&mut self) {
        if !self.initialized {
            return;
        }

        print!(
            "{}{}{}{}",
            escape::RESET_ATTRS,
            escape::MOUSE_OFF,
            escape::CURSOR_SHOW,
            escape::ALT_SCREEN_OFF
        );
        let _ = io::stdout().flush();

        unsafe { libc::tcsetattr(libc::STDIN_FILENO, libc::TCSAFLUSH, &self.orig_termios) };
    }
}
